### PR TITLE
Data Managment: List.SortByKey

### PR DESCRIPTION
**Pull request by aosyatnik**
### Purpose

Link to YouTrack:
[MAGN-8180](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-8180) Provide a method to sort a list by a set of key values

Results:
image

image

image

### Declarations

- [x] The code base is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files

### Reviewers

@mjkkirschner 

### FYIs

@Racel 
@mccrone

### CODE REVIEW

**Comment by mjkkirschner** [comment]
@aosyatnik fix the merge conflicts and I'll take a look :)

File: src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds [hunk 1]
```diff
@@ -458,7 +458,7 @@ def __Reduce(accumulator : _FunctionObject, seed: var[]..[], lists : var[]..[])
     };
 }
 
-def SortByKey(list : var[]..[], keyProjector:_FunctionObject)
+def SortByFunction(list : var[]..[], func:_FunctionObject)
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/ExtendedLibraries/FunctionObject.ds]: looks like this is missing a migration and test

**Comment by aosyatnik** [comment]
Sorry @mjkkirschner , you didn't have to review this until https://github.com/DynamoDS/Dynamo/pull/5260 was not merged... Now 5260 is merged and I can create migration tests. Please, take a look. Thanks!

**Comment by mjkkirschner** [comment]
@aosyatnik theres apparently still a merge conflict on this branch, can you fix it?

**Comment by aosyatnik** [comment]
Sorry, updated one more time. It's up to date with master.

File: src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs [hunk 2]
```diff
@@ -487,6 +487,49 @@ public Filter()
...
+    [NodeName("SortByKey")]
+    [NodeCategory(BuiltinNodeCategories.CORE_LISTS_ACTION)]
+    [NodeDescription("ListSortByKeyDescription", typeof(DSCoreNodesUI.Properties.Resources))]
+    [NodeSearchTags("ListSortByKeySearchTags", typeof(DSCoreNodesUI.Properties.Resources))]
+    [IsDesignScriptCompatible]
+    public class SortByKey : NodeModel
+    {
+        public SortByKey()
+        {
+            InPortData.Add(new PortData("list", ""));
```
Comment (mjkkirschner) [inline src/Libraries/CoreNodeModels/HigherOrder/MapCombineLacing.cs]: no tooltips?

File: test/Engine/ProtoTest/Associative/FunctionObject.cs [hunk 3]
```diff
@@ -594,5 +594,94 @@ public void Test__GroupByKey4()
...
+        public void Test__SortByKey3()
+        {
+            string code =
+    @"
+import (""DSCoreNodes.dll"");
+import (""FunctionObject.ds"");
+
+list = {""item1""};
+keys = {""key1"", ""key2""};
+
```
Comment (mjkkirschner) [inline test/Engine/ProtoTest/Associative/FunctionObject.cs]: can you add a test where the keys are lists and a ping @lukechurch to see if this should work or not.

ie: keys = {{key1},key2,{key1}}.

**Comment by mjkkirschner** [comment]
@aosyatnik looks good :), a small issue with tooltips and another test I think is a good idea.

I think @lukechurch may need to chime in to determine if lists should be keys or not though, if the answer is no then you don't need to add that test.

**Comment by aosyatnik** [comment]
Hi, @mjkkirschner  , test like {{key1},key2,{key1}}  won't pass... I've never thought about case like this. What is expected behavior? Should I turn every {key1} in simply key1? And If we will have something like {{key1,key2}, key1}... What should I do in this case?
@lukechurch , seems we really need your opinion...

**Comment by jnealb** [comment]
@aosyatnik @mjkkirschner @lukechurch For now disallowing lists as keys is fine, imo. Let's see if users ask for it.

**Comment by lukechurch** [comment]
@jnealb Lets see if users ask for it isn't a viable strategy for language design sadly, you end up with an adhoc mess that contains extremely unpleasant surprises :-)

In this case, lists as keys are now allowed, as the list access operators should replicate, so there is no way of accessing them.

Consider e.g.
```
a = {2, 3, 4}
print(a[{0, 1}])
```
The access to the should replicated printing 
```
{2, 3}
```

So the only way to use a list as a key would be the really unplesant O(n) iteration over the keys.

So I'm happy with throwing an "Invalid input data, lists as keys are not supported" exception if you come across that. Good catch @mjkkirschner 

@aosyatnik There seem to be other comments from Mike that you have still to address, so handing this back to you.

**Comment by aosyatnik** [comment]
Hi, @lukechurch , I was able finally run CI. No fallen tests. It can be merged.
image

**Comment by lukechurch** [comment]
lgtm

**Pull request merged**